src/vmm/src/builder.rs

```diff
@@ -148,19 +148,342 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
     }
 }
 
-fn create_vmm_and_vcpus(
+#[cfg(target_arch = "aarch64")]
+pub mod aarch64 {
+    use super::*;
+
+    fn attach_legacy_devices_aarch64(
+        event_manager: &mut EventManager,
+        vmm: &mut Vmm,
+        cmdline: &mut LoaderKernelCmdline,
+    ) -> Result<(), VmmError> {
+        // Serial device setup.
+        let cmdline_contains_console = cmdline
+            .as_cstring()
+            .map_err(|_| VmmError::Cmdline)?
+            .into_string()
+            .map_err(|_| VmmError::Cmdline)?
+            .contains("console=");
+
+        if cmdline_contains_console {
+            // Make stdout non-blocking.
+            set_stdout_nonblocking();
+            let serial = setup_serial_device(event_manager, std::io::stdin(), std::io::stdout())?;
+            vmm.mmio_device_manager
+                .register_mmio_serial(vmm.vm.fd(), &mut vmm.resource_allocator, serial, None)
+                .map_err(VmmError::RegisterMMIODevice)?;
+            vmm.mmio_device_manager
+                .add_mmio_serial_to_cmdline(cmdline)
+                .map_err(VmmError::RegisterMMIODevice)?;
+        }
+
+        let rtc = RTCDevice(Rtc::with_events(
+            &crate::devices::legacy::rtc_pl031::METRICS,
+        ));
+        vmm.mmio_device_manager
+            .register_mmio_rtc(&mut vmm.resource_allocator, rtc, None)
+            .map_err(VmmError::RegisterMMIODevice)
+    }
+
+    /// Configures the system for booting Linux.
+    pub fn configure_system_for_boot(
+        vmm: &mut Vmm,
+        vcpus: &mut [Vcpu],
+        vm_config: &VmConfig,
+        cpu_template: &CustomCpuTemplate,
+        entry_addr: GuestAddress,
+        initrd: &Option<InitrdConfig>,
+        boot_cmdline: LoaderKernelCmdline,
+    ) -> Result<(), StartMicrovmError> {
+        use self::StartMicrovmError::*;
+
+        // Construct the base CpuConfiguration to apply CPU template onto.
+        let cpu_config = {
+            use crate::arch::aarch64::regs::Aarch64RegisterVec;
+            use crate::arch::aarch64::vcpu::get_registers;
+
+            for vcpu in vcpus.iter_mut() {
+                vcpu.kvm_vcpu
+                    .init(&cpu_template.vcpu_features)
+                    .map_err(VmmError::VcpuInit)
+                    .map_err(Internal)?;
+            }
+
+            let mut regs = Aarch64RegisterVec::default();
+            get_registers(&vcpus[0].kvm_vcpu.fd, &cpu_template.reg_list(), &mut regs)
+                .map_err(GuestConfigError)?;
+            CpuConfiguration { regs }
+        };
+
+        // Apply CPU template to the base CpuConfiguration.
+        let cpu_config = CpuConfiguration::apply_template(cpu_config, cpu_template)?;
+
+        let vcpu_config = VcpuConfig {
+            vcpu_count: vm_config.vcpu_count,
+            smt: vm_config.smt,
```

Comment:
hmm.. considering smt does not exist on aarch64 anyway, maybe we can skip it here or always set it to `false`?